unified-scheduler-logic/src/lib.rs

```diff
@@ -562,91 +670,305 @@ impl Default for UsageQueueInner {
+            // queue upon arrival of new higher-prioritized tasks and immediately before the actual
+            // locking (thus this fn is called `prepare_lock()`), _as if those tasks should have
+            // arrived earlier in the precise order of their task_ids. Then, all other code work
+            // nicely with priority ordering enabled.
+            //
+            // Note that reblocking must be consistently applied across all usage queues.
+            // otherwise, deadlock would happen.
+            Self::Priority {
+                current_usage,
+                blocked_usages_from_tasks,
+            } => {
+                // This artificial var is needed to pacify rust borrow checker...
+                let mut current_and_requested_usage = (current_usage, requested_usage);
+
+                match &mut current_and_requested_usage {
+                    (None, _) => {
+                        assert!(blocked_usages_from_tasks.is_empty());
+                        Ok(())
+                    }
+                    (Some(PriorityUsage::Writable(current_task)), _) => {
+                        if !new_task.is_higher_priority(current_task) {
+                            return Err(());
+                        }
+                        if !current_task.try_reblock(token) {
+                            return Err(());
+                        }
+                        let reblocked_task = Usage::take_writable(current_and_requested_usage.0);
+                        blocked_usages_from_tasks
+                            .insert_usage_from_task((RequestedUsage::Writable, reblocked_task));
+                        Ok(())
+                    }
+                    (Some(PriorityUsage::Readonly(_current_tasks)), RequestedUsage::Readonly) => {
+                        match self.peek_usage_from_task() {
+                            None => Ok(()),
+                            Some((usage, peeked_task)) => {
+                                assert_matches!(usage, RequestedUsage::Writable);
+                                if !new_task.is_higher_priority(peeked_task) {
+                                    return Err(());
+                                }
+                                Ok(())
+                            }
+                        }
+                    }
+                    (Some(PriorityUsage::Readonly(current_tasks)), RequestedUsage::Writable) => {
+                        // Use extract_if once stablized to remove Vec creation and the repeating
+                        // remove()s...
+                        let task_indexes = current_tasks
+                            .range(new_task.task_index()..)
+                            .filter_map(|(&index, task)| task.try_reblock(token).then_some(index))
+                            .collect::<Vec<Index>>();
+                        for task_index in task_indexes.into_iter() {
```

Comment:
collect into immediate into_iter. is this necessary to allocate here?

Reply:
i wish we could. but, `current_tasks` couldn't be borrowed immutably by `.range()` and _mutably_ by `.remove()` at the same time. That's why there's years-long desire of still unstable `extract_if`...